go/fn/origin.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package fn
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+const (
+	// upstreamIdentifierRegexPattern provides the rough regex to parse a upstream-identiifier annotation.
+	// "group" should be a domain name. We accept empty string for kubernetes core v1 resources.
+	// "kind" should be the resource type with initial in capitals.
+	// "namespace" should follow RFC 1123 Label Names. We accept "~C~ for cluster-scoped resource or unknown scope resources.
+	// "name" should follow RFC 1123 Label Names https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names
+	upstreamIdentifierRegexPattern = `(?P<group>[a-z0-9-.]*)\|(?P<kind>[A-Z][a-zA-Z0-9]*)\|(?P<namespace>[a-z0-9-]{1,63}|~C)\|(?P<name>[a-z0-9-]{1,63})`
+	upstreamIdentifierFormat       = "<GROUP>|<KIND>|<NAMESPACE>|<NAME>"
+	regexPatternGroup              = "group"
+	regexPatternKind               = "kind"
+	regexPatterNamespace           = "namespace"
+	regexPatternName               = "name"
+)
+
+type ResourceIdentifier struct {
+	Group     string
+	Version   string
+	Kind      string
+	Name      string
+	Namespace string
+}
+
+func (r *ResourceIdentifier) String() string {
+	return fmt.Sprintf("%v|%v|%v|%v", r.Group, r.Kind, r.Namespace, r.Name)
+}
+
+func (o *SubObject) hasUpstreamIdentifier(val interface{}, fields ...string) bool {
```

Comment:
Please add the comment about what is allowed value for `val` and why?

Reply:
Done.